pydantic/_internal/_model_construction.py

```diff
@@ -191,12 +190,7 @@ def wrapped_model_post_init(self: BaseModel, context: Any, /) -> None:
             for name, obj in private_attributes.items():
                 obj.__set_name__(cls, name)
 
-            if __pydantic_reset_parent_namespace__:
-                cls.__pydantic_parent_namespace__ = build_lenient_weakvaluedict(parent_frame_namespace())
-            parent_namespace = getattr(cls, '__pydantic_parent_namespace__', None)
-            if isinstance(parent_namespace, dict):
-                parent_namespace = unpack_lenient_weakvaluedict(parent_namespace)
-
+            parent_namespace = parent_frame_namespace() or {} if __pydantic_reset_parent_namespace__ else {}
             types_namespace = get_cls_types_namespace(cls, parent_namespace)
             set_model_fields(cls, bases, config_wrapper, types_namespace)
```

Comment:
Yet another would be to use the approach in the PR but also make sure the references to parent namespaces are gotten rid of once model has been built. That would also avoid keeping strong references to everything in parent namespaces from models. But would have some risk of references being kept forever if model is not built, or something.

Reply:
@sydney-runkle one more idea would be to add a Config param to allow users skipping the whole parent namespace logic. So when user knows the model doesn't use any forward refs then Config could be used to avoid the overhead with the logic. It would be nice to by default avoid it, but it's too late.